### PR TITLE
Fix Closure issue  when a query expression is used in a let clause of another query expression.

### LINKED ISSUE #34332: Compiler FE crashes when a `let var f =  from/select` clause is used inside a `from/select` expression
**Steps to reproduce:**
The issue can be reproduce by running the following code sample
```ballerina
type Label record {
    string name;
};

type PR record {
    string url;
    string title;
    string state;
    string created_at;
    string updated_at;
    string? merged_at;
    Label[] labels;
};

public function main() returns error? {
    PR[] prs = [
        {
            "url": "https://api.github.com/repos/ballerina-platform/ballerina-lang/pulls/34133",
            "title": "Add missing completions inside ternary conditional expression",
            "state": "closed",
            "created_at": "2021-12-06T04:21:47Z",
            "updated_at": "2021-12-13T11:02:18Z",
            "merged_at": "2021-12-13T11:02:17Z",
            "closed_at": "2021-12-13T11:02:18Z",
            "labels": [
                {
                    "name": "Team/LanguageServer",
                    "default": false,
                    "color": "ead279",
                    "description": "Language Server Implementation related issues. #Compiler",
                    "id": 828273958,
                    "url": "https://api.github.com/repos/ballerina-platform/ballerina-lang/labels/Team/LanguageServer",
                    "node_id": "MDU6TGFiZWw4MjgyNzM5NTg="
                }
            ]
        },
        {
            "url": "https://api.github.com/repos/ballerina-platform/ballerina-lang/pulls/34123",
            "title": "Fix issue in tracking field references in index-based access exprs",
            "state": "closed",
            "created_at": "2021-12-03T16:46:53Z",
            "updated_at": "2021-12-13T06:14:33Z",
            "merged_at": "2021-12-13T05:40:57Z",
            "closed_at": "2021-12-13T05:40:57Z",
            "labels": [
                {
                    "name": "Team/CompilerFETools",
                    "default": false,
                    "color": "ead279",
                    "description": "Semantic API, Formatter, Shell",
                    "id": 1341668468,
                    "url": "https://api.github.com/repos/ballerina-platform/ballerina-lang/labels/Team/CompilerFETools",
                    "node_id": "MDU6TGFiZWwxMzQxNjY4NDY4"
                },
                {
                    "name": "Area/SemanticAPI",
                    "default": false,
                    "color": "186872",
                    "description": "Semantic API Related Issues #Compiler",
                    "id": 2286594184,
                    "url": "https://api.github.com/repos/ballerina-platform/ballerina-lang/labels/Area/SemanticAPI",
                    "node_id": "MDU6TGFiZWwyMjg2NTk0MTg0"
                }
            ]
        }
    ];

    PR[] projectedPRs = from var {url, title, state, created_at, updated_at, merged_at, labels} in prs
        let var projectedLabels = from var {name} in labels
            select {name} // JBUG
        select {url, title, state, created_at, updated_at, merged_at, labels: projectedLabels};

    io:println(projectedPRs);
}
```

**Affected Versions:**
Swan Lake Beta6

### LINKED ISSUE #37809: [Bug]: Fix `incompatible types` diagnostic in query expressions
### Description

Consider the following scenario where the `incompatible types` diagnostic is given in the select clause

https://user-images.githubusercontent.com/61020198/191181224-2b7c2c24-d221-40ee-8733-4856cffafa1c.mov

Instead of giving the diagnostic only to the select clause, it is better to give it to the entire query expression as mentioned in the online discussion. Then the diagnostic message will be `incompatible types: expected 'int', found 'string[]'`

### Steps to Reproduce

```
type Person record {|
    string firstName;
    string lastName;
    int age;
|};

function testFunc() {
    error onConflictError = error("Key Conflict", message = "cannot insert.");
    Person p1 = {firstName: "Alex", lastName: "George", age: 33};
    Person p2 = {firstName: "John", lastName: "David", age: 35};
    Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
    Person[] personList = [p1, p2, p3];

    int outputPersonList =
            from var person in personList
            let int newAge = 20
            where person.age == 33
            select person.firstName;
```

### Related issue(s) (optional)

#36564

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #34332
Fixes #37809

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by gimantha** [comment]
Will send the fix to other branches once the PR is approved

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryNegativeTests.java [hunk 1]
```diff
@@ -104,11 +104,13 @@ public void testFromClauseWithInvalidType() {
         validateError(compileResult, index++, "incompatible types: expected 'error?', found '(error|int)'", 520, 47);
         validateError(compileResult, index++, "incompatible types: expected 'int', found 'string'", 525, 16);
         validateError(compileResult, index++, "incompatible types: expected 'int', found 'string'", 530, 20);
-        Assert.assertEquals(compileResult.getErrorCount(), index);
         validateWarning(compileResult, index++, "invalid usage of the 'check' expression operator:" +
                 " no expression type is equivalent to error type", 544, 15);
         validateWarning(compileResult, index++, "invalid usage of the 'check' expression operator:" +
                 " no expression type is equivalent to error type", 553, 15);
+        validateError(compileResult, index++, "incompatible types: expected 'int', found 'string[]'", 574, 13);
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/QueryNegativeTests.java:111]: Shall we move up the error scenarios in the source and do the error count assertion as well?

**Review by pcnfernando — DISMISSED** [review]
LGTM

**Comment by chiranSachintha** [comment]
Shouldn't this result in a compile-time error? Currently, it is compiling without any issues @gimantha 
```ballerina
type Person record {|
    string firstName;
    string lastName;
    int age;
|};

function testFunc() {
    Person p1 = {firstName: "Alex", lastName: "George", age: 33};
    Person p2 = {firstName: "John", lastName: "David", age: 35};
    Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
    Person[] personList = [p1, p2, p3];

    string outputPersonList =
            from Person person in personList
            let int newAge = 20
            where person.age == 33
            select person.firstName;
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -6200,7 +6200,12 @@ void solveSelectTypeAndResolveType(BLangQueryExpr queryExpr, BLangExpression sel
             case TypeTags.NONE:
             default:
                 // contextually expected type not given (i.e var).
-                selectType = checkExpr(selectExp, env, type, data);
+                selectType = checkExprSilent(nodeCloner.cloneNode(selectExp), type, data);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6203]: Regarding the comment at line 6202, as the default case is being used in the scenarios that the contextually expected type is not provided, don't we need to add additional cases for missing types in the switch statement? If not, can we remove the comment?

Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6203]: Even with the following example, we are still getting errors similar to 'incompatible types: expected 'xml', found 'string
```ballerina
type Person record {|
    string firstName;
    string lastName;
    int age;
|};

function testFunc() {
    Person p1 = {firstName: "Alex", lastName: "George", age: 33};
    Person p2 = {firstName: "John", lastName: "David", age: 35};
    Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
    Person[] personList = [p1, p2, p3];

    xml outputPersonList =
            from Person person in personList
            let int newAge = 20
            where person.age == 33
            select person.firstName;
}
```

**Comment by gimantha** [comment]
> Shouldn't this result in a compile-time error? Currently, it is compiling without any error @gimantha
> 
> ```ballerina
> type Person record {|
>     string firstName;
>     string lastName;
>     int age;
> |};
> 
> function testFunc() {
>     Person p1 = {firstName: "Alex", lastName: "George", age: 33};
>     Person p2 = {firstName: "John", lastName: "David", age: 35};
>     Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
>     Person[] personList = [p1, p2, p3];
> 
>     string outputPersonList =
>             from Person person in personList
>             let int newAge = 20
>             where person.age == 33
>             select person.firstName;
> }
> ```

We can construct a string from a query expression. what error do you expect?

**Comment by chiranSachintha** [comment]
> We can construct a string from a query expression. what error do you expect?

Yeah got it. it can be a String. Based on the error message, I feel that the expected type cannot be a string. 

```ballerina
type Person record {|
    string firstName;
    string lastName;
    int age;
|};

function testFunc() {
    Person p1 = {firstName: "Alex", lastName: "George", age: 33};
    Person p2 = {firstName: "John", lastName: "David", age: 35};
    Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
    Person[] personList = [p1, p2, p3];

    int outputPersonList =
            from Person person in personList
            let int newAge = 20
            where person.age == 33
            select person.firstName; // incompatible types: expected 'int', found 'string[]'
}
```

**Comment by gimantha** [comment]
> > We can construct a string from a query expression. what error do you expect?
> 
> Yeah got it. it can be a String. Based on the error message, I feel that the expected type cannot be a string.
> 
> ```ballerina
> type Person record {|
>     string firstName;
>     string lastName;
>     int age;
> |};
> 
> function testFunc() {
>     Person p1 = {firstName: "Alex", lastName: "George", age: 33};
>     Person p2 = {firstName: "John", lastName: "David", age: 35};
>     Person p3 = {firstName: "Max", lastName: "Gomaz", age: 33};
>     Person[] personList = [p1, p2, p3];
> 
>     int outputPersonList =
>             from Person person in personList
>             let int newAge = 20
>             where person.age == 33
>             select person.firstName; // incompatible types: expected 'int', found 'string[]'
> }
> ```
I remember @KavinduZoysa had an idea to introduce a new error message for this one. Then later we thought we keep the same error message for consistency. But if it affect the user experience, we should change it. @KavinduZoysa WDYT?

**Review by pcnfernando — APPROVED** [review]

**Comment by chiranSachintha** [comment]
Merging this PR would result in inconsistency. In one case, we set the diagnostic only to the select clause, while in another case, we set it for the entire query expression. Please refer to [this](https://github.com/ballerina-platform/ballerina-lang/pull/38390#discussion_r1092760499) review suggestion. And also kindly consider the other review suggestions as well.

**Comment by gimantha** [comment]
> Merging this PR would result in inconsistency. In one case, we set the diagnostic only to the select clause, while in another case, we set it for the entire query expression. Please refer to [this](https://github.com/ballerina-platform/ballerina-lang/pull/38390#discussion_r1092760499) review suggestion. And also kindly consider the other review suggestions as well.

@chiranSachintha Missed it :(, Will fix it in a different PR

**Pull request merged**